### PR TITLE
Replace @overload_glue by @overload for 20 NumPy functions

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Turn the functions below into `@overload` functions. To make things a bit easier, I'm listing the changes I did for each one of the functions.

### Kind: Construct Functions

**Functions:**

- [x]  `np.array(object, dtype=None)`
    - Raise `TypingError` if `object` is not array-like
    - Raise `TypingError` if cannot parse the dtype.
- [x]  `np.diag(v, k=0)`
    - `test_dyn_array::TestNdDiag` already checks for some typing errors
    - Raise `TypingError` if `v` is not array-like
- [x]  `np.full(shape, fill_value, dtype=None)`
    - Changed `test_1d_dtype_invalid_str` input from `np.ones(4)` to `(5,)`. The previous input is not valid in NumPy
    - Raise `TypingError` if cannot parse dtype
- [x]  `np.full_like(a, fill_value, dtype=None)`
    - Raise `TypingError` if `a` is not array-like
    - Raise `TypingError` if cannot parse the dtype.

---

### Kind: Routine Functions

- [x]  `np.ascontiguousarray(a)`
    - Update `numpysupported.rst` to include it in the list of supported NumPy functions
    - Raise `TypingError` if `a` is not array-like
- [x]  `np.asfortranarray(a)`
    - Raise `TypingError` if `a` is not array-like
- [x]  `np.copy(a)`
    - Raise `TypingError` if `a` is not array-like
- [x]  `np.expand_dims(a, axis)`
    - Added tests to validate types `a: array-like` and `axis: int|Tuple[int]`
- [x]  `np.frombuffer(buffer, dtype=float)`
    - Raise `TypingError` if `buffer` is not buffer-like
- [x]  `np.identity(n, dtype=None)`
    - Return `None` if dtype is a `UnicodeType` rather than raise a `RequireLiteralValue` error
- [x]  `np.round/np.around(a, decimals=0, out=None)`
    - Added tests to validate types of `np.round/np.around` in `test_np_functions.py`
    - Added tests to validate types `a: array_like` and `out: Optional[array]`
    - Update `numpysupported.rst` to include `numpy.around` in the list of supported NumPy functions
- [x]  `np.concatenate(arrays, axis=0)`
    - Added tests to validate `array: Tuple[array-like]`
- [x]  `np.stack(arrays, axis=0)`
    - function `sequence_of_arrays` checks if `arrays` is a Tuple[array-like]. Include test for this case. This is also valid for `np.hstack/vstack/dstack/column_stack`
    - Update `numpysupported.rst` to include only the first two arguments are supported
- [x]  `np.column_stack(tup)`
- [x]  `np.dstack(tup)`
- [x]  `np.hstack(tup)`
- [x]  `np.vstack(tup)`

---

### Kind: Sorting

**functions:**

- [x]  `np.sort(a)`
    - Raise `TypingError` if `a` is not array-like

---

### Kind: Mathematical Functions

**functions:**

- [x]  `np.angle`
    - Update test `test_errorhandling.py::TestErrorMessage.test_callable_template_source`
    - Add test to validate types on `test_np_functions.py`
- [x]  `np.sinc(x)`
    - Raise `TypeError` if `x` is not number or array-like

### CODE REVIEW

File: numba/np/arrayobj.py [hunk 1]
```diff
@@ -4731,45 +4705,80 @@ def array_astype(context, builder, sig, args):
     return impl_ret_new_ref(context, builder, sig.return_type, ret._getvalue())
 
 
-@glue_lowering(np.frombuffer, types.Buffer)
-@glue_lowering(np.frombuffer, types.Buffer, types.DTypeSpec)
-@glue_lowering(np.frombuffer, types.Buffer, types.StringLiteral)
-def np_frombuffer(context, builder, sig, args):
-    bufty = sig.args[0]
-    aryty = sig.return_type
+# @glue_lowering(np.frombuffer, types.Buffer)
```
Comment (apmasell) [inline numba/np/arrayobj.py:4708]: What's going on with these?

File: numba/tests/test_array_methods.py [hunk 2]
```diff
@@ -409,13 +423,13 @@ def check_err(arr, dtype):
...
 
         # F-contiguous
-        arr = np.arange(24, dtype=np.int8).reshape((3, 8)).T
-        check(arr, np.int8)
-        check(arr, np.float32)
-        check(arr, np.complex64)
-        check(arr, dt1)
-        check_err(arr, dt2)
-        check_err(arr, np.complex128)
+        # arr = np.arange(24, dtype=np.int8).reshape((3, 8)).T
```
Comment (apmasell) [inline numba/tests/test_array_methods.py:426]: Should these be enabled?

**Review by apmasell — COMMENTED** [review]
The random number code has a lot of similar work to handle generating arrays of a particular shape. I think, in the future, we should try to refactor that code to reduce duplication.

File: numba/tests/test_array_methods.py (hunk 2 above)
Comment (guilhermeleobas) [reply numba/tests/test_array_methods.py:426]: Done!

File: numba/np/arrayobj.py (hunk 1 above)
Comment (guilhermeleobas) [reply numba/np/arrayobj.py]: Oops, forgot to remove them

**Review by apmasell — APPROVED** [review]

**Review by apmasell — DISMISSED** [review]

**Comment by sklam** [comment]
@guilhermeleobas, there's a conflict with mainline

**Comment by guilhermeleobas** [comment]
> @guilhermeleobas, there's a conflict with mainline

Fixed!

**Review by sklam — APPROVED** [review]

**Pull request merged**